From 826acdfce86a82284cf512731b798be3ad6b3138 Mon Sep 17 00:00:00 2001
From: Sami Shalayel <sami.shalayel@qt.io>
Date: Thu, 16 Oct 2025 15:37:26 +0200
Subject: [PATCH 191/239] qmllint: prefer enums over chained attached type

It seems that FlexboxLayout.Row is actually ambiguous: it could refer to
the "Row" enum value, and to the attached type of "Row" which is
attached on the attached type of "FlexboxLayout".

Fix qmllint to follow the qml engine's interpretation of the thing, and
resolve it as an enum instead of a chained attached type in
QQmlJSTypeResolver.

Pick-to: 6.8
Fixes: QTBUG-141194
Change-Id: I24e23f5fc92b0d007ff1d628b6286f85a60f10d8
Reviewed-by: Fabian Kosmale <fabian.kosmale@qt.io>
(cherry picked from commit adb6b6861b295d0f99746428d02fce4cff5cdc1e)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/qmlcompiler/qqmljstyperesolver.cpp        | 19 +++---
 .../qmllint/data/EnumList/enumlist.qmltypes   | 62 +++++++++++++++++--
 tests/auto/qml/qmllint/tst_qmllint.cpp        |  6 +-
 3 files changed, 71 insertions(+), 16 deletions(-)

diff --git a/src/qmlcompiler/qqmljstyperesolver.cpp b/src/qmlcompiler/qqmljstyperesolver.cpp
index 555eecc9e0..2e9b268204 100644
--- a/src/qmlcompiler/qqmljstyperesolver.cpp
+++ b/src/qmlcompiler/qqmljstyperesolver.cpp
@@ -1557,7 +1557,7 @@ QQmlJSRegisterContent QQmlJSTypeResolver::memberType(
 
     // If we got a plain type reference we have to check the enums of the _scope_.
     if (contained == metaObjectType())
-        return {};
+        return memberEnumType(type.scope(), name);
 
     if (contained == variantMapType() || contained->inherits(qmlPropertyMapType())) {
         QQmlJSMetaProperty prop;
@@ -1633,6 +1633,12 @@ QQmlJSRegisterContent QQmlJSTypeResolver::memberType(
         }
     }
 
+    // check enums before checking attached types of attached types (chained attached types)
+    if (type.isType()) {
+        if (auto result = memberEnumType(type.scope(), name); result.isValid())
+            return result;
+    }
+
     if (QQmlJSScope::ConstPtr attachedBase = typeForName(name)) {
         if (QQmlJSScope::ConstPtr attached = attachedBase->attachedType()) {
             if (!genericType(attached)) {
@@ -1682,15 +1688,8 @@ QQmlJSRegisterContent QQmlJSTypeResolver::memberEnumType(
 QQmlJSRegisterContent QQmlJSTypeResolver::memberType(
         QQmlJSRegisterContent type, const QString &name, int lookupIndex) const
 {
-    if (type.isType()) {
-        const auto result = memberType(type, name, type.resultLookupIndex(), lookupIndex);
-        if (result.isValid())
-            return result;
-
-        // If we didn't find anything and it's an attached type,
-        // we might have an enum of the attaching type.
-        return memberEnumType(type.scope(), name);
-    }
+    if (type.isType())
+        return memberType(type, name, type.resultLookupIndex(), lookupIndex);
     if (type.isProperty() || type.isMethodCall())
         return memberType(type, name, type.resultLookupIndex(), lookupIndex);
     if (type.isEnumeration()) {
diff --git a/tests/auto/qml/qmllint/data/EnumList/enumlist.qmltypes b/tests/auto/qml/qmllint/data/EnumList/enumlist.qmltypes
index 031f7406f6..c776ac09e2 100644
--- a/tests/auto/qml/qmllint/data/EnumList/enumlist.qmltypes
+++ b/tests/auto/qml/qmllint/data/EnumList/enumlist.qmltypes
@@ -1,10 +1,5 @@
 import QtQuick.tooling 1.2
 
-// This file describes the plugin-supplied types contained in the library.
-// It is used for QML tooling purposes only.
-//
-// This file was auto-generated by qmltyperegistrar.
-
 Module {
     Component {
         file: "mycustomtype.h"
@@ -48,4 +43,61 @@ Module {
             values: ["A", "B", "C", "D"]
         }
     }
+    Component {
+        file: "private/qquickflexboxlayout_p.h"
+        name: "QQuickFlexboxLayout"
+        accessSemantics: "reference"
+        prototype: "QObject"
+        exports: ["EnumProperty/FlexboxLayout 1.0"]
+        exportMetaObjectRevisions: [256]
+        attachedType: "QQuickFlexboxLayoutAttached"
+        Enum {
+            name: "FlexboxDirection"
+            values: ["Column", "ColumnReverse", "Row", "RowReverse"]
+        }
+        Property {
+            name: "direction"
+            type: "FlexboxDirection"
+            read: "direction"
+            write: "setDirection"
+            notify: "directionChanged"
+            index: 0
+            isFinal: true
+        }
+    }
+    Component {
+        file: "private/qquickflexboxlayout_p.h"
+        name: "QQuickFlexboxLayoutAttached"
+        accessSemantics: "reference"
+        prototype: "QObject"
+    }
+    Component {
+        file: "private/qquickpositioners_p.h"
+        name: "QQuickRow"
+        accessSemantics: "reference"
+        prototype: "QObject"
+        attachedType: "QQuickPositionerAttached"
+        exports: [
+            "EnumProperty/Row 1.0",
+        ]
+        exportMetaObjectRevisions: [
+            256,
+        ]
+    }
+    Component {
+        file: "private/qquickpositioners_p.h"
+        name: "QQuickPositionerAttached"
+        accessSemantics: "reference"
+        prototype: "QObject"
+        Property {
+            name: "propertyOfAttachedTypeRow"
+            type: "int"
+            read: "propertyOfAttachedTypeRow"
+            write: "setPropertyOfAttachedTypeRow"
+            notify: "propertyOfAttachedTypeRowChanged"
+            index: 0
+            isFinal: true
+        }
+    }
+
 }
diff --git a/tests/auto/qml/qmllint/tst_qmllint.cpp b/tests/auto/qml/qmllint/tst_qmllint.cpp
index 3cf73d6d86..f284e15fe3 100644
--- a/tests/auto/qml/qmllint/tst_qmllint.cpp
+++ b/tests/auto/qml/qmllint/tst_qmllint.cpp
@@ -1581,6 +1581,9 @@ void TestQmllint::cleanQmlSnippet_data()
     QTest::newRow("usefulExpressionStatement") << u"x: y + 3;"_s << defaultOptions;
     QTest::newRow("usefulExpressionStatement") << u"x: 3;"_s << defaultOptions;
     QTest::newRow("void") << u"function f(): void {}"_s << defaultOptions;
+    QTest::newRow("ambiguity-enum-and-chained-attached-property")
+            << u"import EnumList\nFlexboxLayout { direction: FlexboxLayout.Row; }"_s
+            << defaultOptions;
 }
 
 void TestQmllint::cleanQmlSnippet()
@@ -1588,7 +1591,8 @@ void TestQmllint::cleanQmlSnippet()
     QFETCH(QString, code);
     QFETCH(CallQmllintOptions, options);
 
-    const QString qmlCode = "import QtQuick\nItem {%1}"_L1.arg(code);
+    const QString qmlCode =
+            code.startsWith("import"_L1) ? code : "import QtQuick\nItem {%1}"_L1.arg(code);
     const Result result = Result::clean();
 
     const QJsonArray warnings =
-- 
2.51.2

